src/pages/files.tsx

```diff
@@ -1,64 +1,275 @@
+              format: isImage ? 'base64' : 'text',
+              content: finalContent
+            });
+          } catch (error) {
+            console.warn(`Upload skipped or failed for ${finalFileName}`, error);
+          }
+        }
+      } finally {
+        props.onUploadEnd();
+      }
+    },
+    [props]
+  );
+
+  const handleInputChange = (e: React.ChangeEvent<HTMLInputElement>) => {
+    if (e.target.files) {
+      handleFileSelect(e.target.files);
+    }
+  };
+
+  const triggerFileSelect = () => {
+    fileInputRef.current?.click();
+  };
+
+  // Expose the trigger function to the parent
+  React.useImperativeHandle(ref, () => ({
+    triggerFileSelect
+  }));
+
+  return (
+    <input
+      ref={fileInputRef}
+      type="file"
+      multiple
+      onChange={handleInputChange}
+      style={{ display: 'none' }}
+      accept=".png,.jpg,.jpeg,.csv,image/png,image/jpeg,text/csv"
+    />
+  );
+});
+
+FileUploader.displayName = 'FileUploader';
+
+/**
+ * The main Files page component. It manages the state of uploaded files,
+ * handles file uploads, and renders the file thumbnails.
+ */
+interface IFilesAppProps {
+  contentsManager: Contents.IManager;
+}
+
+function FilesApp(props: IFilesAppProps) {
+  const [listing, setListing] = useState<Contents.IModel | null>(null);
+  const [isUploading, setIsUploading] = useState(false);
+  const fileUploaderRef = useRef<IFileUploaderRef>(null);
+
+  const refreshListing = useCallback(async () => {
+    try {
+      const dirListing = await props.contentsManager.get('', { content: true });
+      setListing(dirListing);
+    } catch (err) {
+      await showErrorMessage(
+        'Error loading files',
+        `Could not load files from the contents manager: ${
+          err instanceof Error ? err.message : String(err)
+        }`
+      );
+    }
+  }, [props.contentsManager]);
+
+  useEffect(() => {
+    void refreshListing();
+  }, [refreshListing]);
+
   return (
-    <button className="je-Tile">
-      <div className="je-Tile-icon">
-        <props.icon.react />
+    <div className="je-FilesApp">
+      <FileUploader
+        ref={fileUploaderRef}
+        onUploadStart={() => setIsUploading(true)}
+        onUploadEnd={async () => {
+          setIsUploading(false);
+          await refreshListing();
+        }}
+        contentsManager={props.contentsManager}
+      />
+      <div className="je-FilesApp-content">
+        <div className="je-FilesApp-grid">
+          {/* "add new" tile */}
+          <div className="je-FileTile">
+            <div
+              className="je-FileTile-box"
+              onClick={() => fileUploaderRef.current?.triggerFileSelect()}
+            >
+              {isUploading ? (
+                <div className="je-FileTile-spinner" />
+              ) : (
+                <EverywhereIcons.addFile.react />
+              )}
+            </div>
+            <div className="je-FileTile-label">{isUploading ? 'Uploading file...' : 'add new'}</div>
```

Comment:
This loading indicator is not included in the Figma design documents right now, but I thought it makes sense, as it is possible to upload more than one file or a large one, which can take time.